gapis/memory/pool.go

```diff
@@ -62,65 +69,223 @@ const (
-func (m *Pool) Slice(rng Range) Data {
-	i, c := interval.Intersect(&m.writes, rng.Span())
-	if c == 1 {
-		w := m.writes[i]
-		if rng == w.dst {
-			// Exact hit
-			return w.src
+// ReserveRanges makes a set of ranges available in the
+// application pool.
+func (m *Pool) ReserveRanges(ranges interval.U64RangeList) {
+	if m.id != ApplicationPool {
+		panic("You can only reserve ranges in the application pool")
+	}
+	for idx := range ranges {
+		alloc := allocation{
+			ranges[idx].First,
+			ranges[idx].Count,
+			m.arena.Allocate(uint32(ranges[idx].Count), 1),
+		}
+		m.allocations = insertAllocation(m.allocations, alloc)
+	}
+}
+
+// AddRange makes a single allocation available to the application
+// pool
+func (m *Pool) AddRange(offset, count uint64) {
+	if m.id != ApplicationPool {
+		panic("You can only reserve ranges in the application pool")
+	}
+
+	alloc := allocation{
+		offset,
+		count,
+		m.arena.Allocate(uint32(count), 1),
+	}
+	m.allocations = insertAllocation(m.allocations, alloc)
+}
+
+// RemoveRange removes a single allocation from the application pool.
+// It is invalid to remove an allocation that was never added
+func (m *Pool) RemoveRange(offset uint64) {
+	if m.id != ApplicationPool {
+		panic("You can only remove ranges from the application pool")
+	}
+	i := sort.Search(len(m.allocations), func(i int) bool { return m.allocations[i].offset >= offset })
+	if i >= len(m.allocations) || m.allocations[i].offset != offset {
+		msg := fmt.Sprintf("Removed allocation [%d] was not found in arena %+v", offset, m.allocations)
+		panic(msg)
+	}
+	m.arena.Free(m.allocations[i].data)
+	copy(m.allocations[i:], m.allocations[i+1:])
+	m.allocations = m.allocations[:len(m.allocations)-1]
+}
+
+func (m *Pool) tryGetAllocation(base uint64) int {
+	i := sort.Search(len(m.allocations), func(i int) bool { return m.allocations[i].offset > base })
+	if i == 0 {
+		return len(m.allocations) + 1
+	}
+
+	if i >= len(m.allocations) && 0 != len(m.allocations) {
+		i = len(m.allocations)
+	}
+
+	return i - 1
+}
+
+func (m *Pool) mustGetAllocation(base uint64) int {
+	i := sort.Search(len(m.allocations), func(i int) bool { return m.allocations[i].offset > base })
+	if i == 0 {
+		panic(fmt.Sprintf("Could not find allocation in pool: %+v, %+v", base, m.allocations))
+	}
+
+	if i >= len(m.allocations) && 0 != len(m.allocations) {
+		i = len(m.allocations)
+	}
+
+	i--
+	allocation := m.allocations[i]
+	if base > allocation.offset+allocation.size {
+		panic(fmt.Sprintf("Could not find allocation in pool: %+v, %+v, %+v", base, allocation, m.allocations))
+	}
+	return i
+}
+
+func (m *Pool) mustGetOffsetAndAllocation(rng Range) (int, uint64) {
+	i := m.mustGetAllocation(rng.Base)
+	allocation := m.allocations[i]
+	if allocation.offset+allocation.size < rng.Base+rng.Size {
+		panic(fmt.Sprintf("Slicing outside of range %+v %+v %+v \n\n---------------------------\n%+v\n----------------------\n", rng, rng.Size, allocation, m.allocations))
```

Comment:
We probably want to make this a bit of a nicer message, it is useful for debug, but prints out EVERY valid range in the pool.